infrastructure/crypto/src/signatures.rs

```diff
@@ -2,25 +2,98 @@
 //! This module defines generic traits for handling the digital signature operations, agnostic
 //! of the underlying elliptic curve implementation
 
-use crate::keys::{PublicKey, SecretKey};
+use crate::{
+    challenge::Challenge,
+    keys::{PublicKey, SecretKey},
+};
+use derive_error::Error;
+use digest::Digest;
+use std::ops::{Add, Mul};
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum SchnorrSignatureError {
+    // An invalid challenge was provided
+    InvalidChallenge,
+}
 
-/// Generic definition of Schnorr Signature functionality, agnostic of the elliptic curve used.
-/// Schnorr signatures are linear and have the form _s = r + ek_, where _r_ is a nonce (secret key),
-/// _k_ is a secret key, and _s_ is the signature.
 #[allow(non_snake_case)]
-pub trait SchnorrSignature {
-    type Scalar: SecretKey;
-    type Point: PublicKey;
-    type Challenge;
+#[derive(PartialEq, Eq, Copy, Debug, Clone)]
+pub struct SchnorrSignature<P, K>
+where
+    P: PublicKey<K = K>,
+    K: SecretKey,
+{
+    public_nonce: P,
+    signature: K,
+}
+
+impl<P, K> SchnorrSignature<P, K>
+where
+    P: PublicKey<K = K>,
+    K: SecretKey,
+{
+    pub fn new(public_nonce: P, signature: K) -> Self {
+        SchnorrSignature { public_nonce, signature }
+    }
 
-    fn new(public_nonce: Self::Point, signature: Self::Scalar) -> Self;
+    pub fn calc_signature_verifier(&self) -> P {
+        P::from_secret_key(&self.signature)
+    }
 
-    fn sign(secret: &Self::Scalar, nonce: &Self::Scalar, challenge: Self::Challenge) -> Self;
+    pub fn sign<'a, 'b, D: Digest>(
+        secret: K,
+        nonce: K,
+        challenge: Challenge<D>,
+    ) -> Result<Self, SchnorrSignatureError>
+    where
+        K: Add<Output = K> + Mul<P, Output = P> + Mul<Output = K>,
+    {
+        // s = r + e.k
+        let e = match K::from_vec(&challenge.hash()) {
+            Ok(e) => e,
+            Err(_) => return Err(SchnorrSignatureError::InvalidChallenge),
+        };
+        let public_nonce = P::from_secret_key(&nonce);
+        let ek = e * secret;
+        let s = ek + nonce;
+        Ok(Self::new(public_nonce, s))
+    }
 
-    /// Check whether the given signature is valid for the given message and public key
-    fn verify(&self, public_key: &Self::Point, challenge: &Self::Challenge) -> bool;
+    pub fn verify<'a, D: Digest>(&self, public_key: &'a P, challenge: Challenge<D>) -> bool
```

Comment:
Challenge here is can be easily replace by just a plain `Digest`? Instead of `.hash()` you call `.result()`?

Reply:
Is there a reason we use challenge in the first place and not just straight digest?

Reply:
Yeah, it's a pretty thin wrapper. The idea was to provide a nice fluent interface; and implement ByteArray on Challenge (though the latter idea hasn't panned out)